### PR TITLE
Generate Server Code

### LINKED ISSUE #82: Generate a short server code when study is created
* Server code is too long
* Make a separate field on studies to store the server code and generate a random one when the study is created
* Show this server code instead of the study id at the top of the study builder

### PR DESCRIPTION
**Pull request by amberchow8**
-Added short-unique-id library 
-Added middleware to Study schema where server code is generated when a Study is created in the database
-Changed server code to serverCode property of the Study on the StudyBuilderPage
-Closes #82 

*I couldn't figure out how to update current studies in the database with the new serverCode property but I created new studies on the UI and their server codes show up

### CODE REVIEW

File: packages/api/src/models/study.ts [hunk 1]
```diff
@@ -36,6 +40,13 @@ const studySchema = new Schema<IStudy>({
   description: { type: String, default: "" },
   batteries: [{ type: Schema.Types.ObjectId, ref: "CustomizedBattery" }],
   sessions: [sessionSchema],
+  serverCode: { type: String, unique: true },
 });
 
+studySchema.pre("save", async function (next) {
+  if (!this.serverCode) {
+    this.serverCode = uid.rnd(5);
```
Comment (amberchow8) [inline packages/api/src/models/study.ts:48]: Not sure if length of 5 is long enough

File: packages/ui/src/stores/studyBuilder.ts [hunk 2]
```diff
@@ -192,6 +195,7 @@ export const useStudyBuilderStore = defineStore("studyBuilder", () => {
       description: description.value ?? "",
       batteries: taskBank.value.map((id) => taskData.value[id]), // TODO: fix this
       sessions: sessions.value.map((id) => sessionData.value[id]),
+      serverCode: serverCode.value!,
```
Comment (ross3102) [inline packages/ui/src/stores/studyBuilder.ts:198]: can this be undefined? curious why the `!`

**Review by ross3102 — APPROVED** [review]
looks good! just small comments

**Comment by amberchow8** [comment]
I changed the pre middleware function a little to regenerate the server code if the first one generated is found in a study. I also added the alphanumeric lowercase dictionary for code generation so all letters are lowercase.

File: packages/api/src/models/study.ts [hunk 3]
```diff
@@ -36,6 +41,18 @@ const studySchema = new Schema<IStudy>({
...
+  serverCode: { type: String, unique: true },
 });
 
+studySchema.pre("save", async function (next) {
+  if (!this.serverCode) {
+    this.serverCode = uid.rnd(serverCodeLength);
+
+    const study = await Study.findOne({ serverCode: this.serverCode });
+    if (study) {
+      this.serverCode = uid.rnd(serverCodeLength);
```
Comment (ross3102) [inline packages/api/src/models/study.ts:55]: good approach, but this will need to be a while loop of some kind in case the regenerated code is also taken

**Review by ross3102 — CHANGES_REQUESTED** [review]
all good except making sure the code generation never overlaps!

**Review by ross3102 — APPROVED** [review]
LGTM!

**Pull request merged**